### ISSUE TITLE
Duda sobre resolución parte de herencia en DER/MAPEO parcial Ginpass

### ISSUE THREAD

**Issue by CamiMargni** [dds-jv/foro#77]
Hola chicos, dejo el der y las aclaraciones del enunciado
![image](https://user-images.githubusercontent.com/48899516/99918392-57686080-2cf5-11eb-8390-e6973485a87d.png)
![image](https://user-images.githubusercontent.com/48899516/99918400-5fc09b80-2cf5-11eb-9412-bc7a5c602068.png)
 *dejo el link por las dudas https://docs.google.com/document/d/18uB4Bb5qnfsrvH3I6sDWCF6LxlBvY7mHkFkTBrRdOG4/edit#
Mi duda es primero para validar una elección de tipo de herencia y luego para ver si en la otra hay algún formato sustancialmente mejor a otro porque estoy entre 2.
Respecto a la herencia en local mi elección y justificación es:
Dado que no se aclara ni en el diagrama de clases ni en las aclaraciones debajo si un local se consulta por su tipo, ej: sólo bares, sólo cervecerías se puede asumir que no es necesario consultarlos individualmente, no polimórficamente, en cuanto a las suscripciones se ve que agrupa locales por los costos de las bebidas y no por sus tipos. Por lo cual se podría descartar la forma de herencia de tipo “por clase concreta” ya que su ventaja allí es la consulta de los subtipos de forma separada que acá no es requerida, debería hacerse siempre que se filtren locales joins para cada una de las 3 subtablas que podrían evitarse con por ejemplo una sóla tabla.  Además el manejo de ids debería hacerse por ejemplo en una tabla aparte y si se insertan varios locales (que podría suceder ya que se aclara la alta oferta que está surgiendo en la ciudad) esa tabla generaría un cuello de botella. 
* Aclaración: Asumo que está mal diagramado el tema de tragos tipo cervezas y cervezas de tipo trago en bares. Es al revés.
Además de repetir 6 columnas siempre en cada una de las tablas (las que corresponden a la superclase local nombre y las correspondientes a dirección), por otro lado, quedan a posible elección la tabla única o el formato joined/tabla por clase. 
![image](https://user-images.githubusercontent.com/48899516/99918443-98f90b80-2cf5-11eb-832c-104aefb83c0e.png)
Como los demás atributos corresponden a listas de elementos de tipo many to many (ej: una cervecería puede tener muchas cervezas, y una cerveza puede estar en muchas cervecerías) no habría campos que difieren entre cada subclase. El mejor caso entonces es una tabla sola con un discriminador. Mejor para las consultas polimórficas, sería un select sencillo y si se quisiera filtrar sería un where al discriminador y no left joins como en un tipo joined. No habría columnas nulas.

Está bien esa deducción y elección?

Y por otro lado, para lo que es bebidas:
Para evaluar en bebida que formato de herencia tomar es importante determinar varias cosas como ventaja y desventaja:
Se consulta en la búsqueda y no se filtran por tipos de bebida (trago, cerveza o sin alcohol) sino por su nombre, por ese lado podría usarse una sola tabla que mejore performance al consultar polimórficamente, también,al ser la bebida, en nuestro dominio sólo un string con el nombre y no tiene otros atributos relevantes como componentes, valores numéricos u otros no lo hace lo suficientemente relevante para crear una tabla sola de “bebida” que deba implementar siempre un join a al menos una subclase si se quisiera acceder no polimórficamente a sus hijos., además todas sus subclases menos la cerveza tampoco contienen atributos distintos entre sí. Pero, en contraparte el enunciado aclara que la cerveza siempre debe tener el tipo, si se usara una sola tabla el control de no nulos no sería factible, ya que si es una bebida sin alcohol la columna de tipo de cerveza debería ser si o si nula.Lo que sí puede estar presente en un tipo Joined con restricciones acordes. Entonces, veo más ventajoso el uso de una tabla, siempre validando del lado de objetos que la cerveza a cargar tenga tipo. Pero si se quisiera un mayor control castigando un poco más la performance se podría ir por un formato joined. A efectos del parcial y para diferenciar elijo formato join

Dudas puntuales:
1.Qué significa en aclaraciones que la lista de bebidas elegidas por un usuario no sea consultada más tarde? influye en algo de lo dicho anteriormente?
2. Si aclara que si o si debe estar el tipo de cerveza, se puede justificar tomando que se valide desde objetos esa carga? o debe ser si o si validado también desde el modelo relacional con restricciones? porque veo más ventajoso por performance y consultas tipo join o group by en bebidas la tabla única aunque no puedan aplicarse dichas restricciones.

**Comment by gastonprieto** [comment]
> Dado que no se aclara ni en el diagrama de clases ni en las aclaraciones debajo si un local se consulta por su tipo, ej: sólo bares, sólo cervecerías se puede asumir que no es necesario consultarlos individualmente, no polimórficamente, 

Correcto

> en cuanto a las suscripciones se ve que agrupa locales por los costos de las bebidas y no por sus tipos. 
Acá no entendí a que te referís por el costo de las bebidas, pero si la suscripción agrupa locales.

> Por lo cual se podría descartar la forma de herencia de tipo “por clase concreta” ya que su ventaja allí es la consulta de los subtipos de forma separada que acá no es requerida, debería hacerse siempre que se filtren locales joins para cada una de las 3 subtablas que podrían evitarse con por ejemplo una sóla tabla. Además el manejo de ids debería hacerse por ejemplo en una tabla aparte y si se insertan varios locales (que podría suceder ya que se aclara la alta oferta que está surgiendo en la ciudad) esa tabla generaría un cuello de botella. 

👍 

> Además de repetir 6 columnas siempre en cada una de las tablas (las que corresponden a la superclase local nombre y las correspondientes a dirección)

No solo eso, van a ser las únicas columnas que tengan, no tienen mas columnas que sean distintas (creo que un poco explicas esto en lo que sigue)

> Como los demás atributos corresponden a listas de elementos de tipo many to many (ej: una cervecería puede tener muchas cervezas, y una cerveza puede estar en muchas cervecerías) no habría campos que difieren entre cada subclase. El mejor caso entonces es una tabla sola con un discriminador. Mejor para las consultas polimórficas, sería un select sencillo y si se quisiera filtrar sería un where al discriminador y no left joins como en un tipo joined. No habría columnas nulas.
Está bien esa deducción y elección?

Genial 👍 

>Se consulta en la búsqueda y no se filtran por tipos de bebida (trago, cerveza o sin alcohol) sino por su nombre, por ese lado podría usarse una sola tabla que mejore performance al consultar polimórficamente, también,al ser la bebida, en nuestro dominio sólo un string con el nombre y no tiene otros atributos relevantes como componentes, valores numéricos u otros no lo hace lo suficientemente relevante para crear una tabla sola de “bebida” que deba implementar siempre un join a al menos una subclase si se quisiera acceder no polimórficamente a sus hijos., además todas sus subclases menos la cerveza tampoco contienen atributos distintos entre sí. 

Correcto

> Pero, en contraparte el enunciado aclara que la cerveza siempre debe tener el tipo, si se usara una sola tabla el control de no nulos no sería factible, ya que si es una bebida sin alcohol la columna de tipo de cerveza debería ser si o si nula.Lo que sí puede estar presente en un tipo Joined con restricciones acordes. Entonces, veo más ventajoso el uso de una tabla, siempre validando del lado de objetos que la cerveza a cargar tenga tipo. Pero si se quisiera un mayor control castigando un poco más la performance se podría ir por un formato joined

Genial !

_Por fuera de la materia: Hay mecanismos como para hacer esta validación por ejemplo con un trigger cuando esten tratando de hacerte un `insert/update` pero esto escapa a la materia_

> . A efectos del parcial y para diferenciar elijo formato join

Hubiera elegido igual la otra 😋  , pero igual como te diste cuenta de las opciones esta ok !

> 1.Qué significa en aclaraciones que la lista de bebidas elegidas por un usuario no sea consultada más tarde? influye en algo de lo dicho anteriormente?

Qué no hace falta persistir el objeto **Busqueda**

> 2. Si aclara que si o si debe estar el tipo de cerveza, se puede justificar tomando que se valide desde objetos esa carga? o debe ser si o si validado también desde el modelo relacional con restricciones? porque veo más ventajoso por performance y consultas (tipo join o group by) en bebidas la tabla única aunque no puedan aplicarse dichas restricciones.

Alcanza con decir la problematica, las opciones y si queres podes quedarte con la de objetos .

**Comment by gastonprieto** [comment]
Una cosa mas importante de este parcial es como quedaron las manytomany entre Bebidas y Locales.

**Comment by CamiMargni** [comment]
![image](https://user-images.githubusercontent.com/48899516/100019618-1b98ce00-2dbd-11eb-90de-36c1cc79c3da.png)
 Yo lo modelé así

**Comment by gastonprieto** [comment]
Buenas @CamiMargni !

En este caso serían 3 tablas distintas, no una `BebidaBar`, porque son 3 relaciones distintas, sino cuando cree la consulta para saber las entidades no va a saber con cual de todas corresponde. En general cada relacion de objetos tiene su propia relacion en objetos. Por ejemplo: Si un objeto tiene dos atributos hacia algo del mismo tipo, debería tener 2 FKs, cada una con su correspondiente relación, acá es lo mismo pero con muchos a muchos

**Comment by FeliIsra** [comment]
Hola @gastonprieto ! Entonces quedarian 3 tablas intermedias o como ? No me queda claro.

(Yo lo habia modelado como arriba tambien)

**Comment by gastonprieto** [comment]
Exacto, son 5 tablas. y tienen que usar el [JoinTable](https://docs.jboss.org/hibernate/jpa/2.1/api/javax/persistence/JoinTable.html) para que esto ande, en cada relación le meten el nombre que corresponde 

Una por cada atributo ManyToMany

**Comment by tatonico** [comment]
Buenas @gastonprieto , yo lo pense como los chicos de arriba y la verdad que le encontraba todo el sentido al DER y al modelo de objetos, no comprendo qué es lo que te limita usar una sola tabla intermedia si a fin y al cabo eso se delega en el ORM y basta con tener un ID de ambos lados como para saber a que registro especifico de cada tabla refiere. Perdon por la molestia, pero ¿podrias especificar un toque mas a que te referis con que no podria saber a cual corresponde?

**Comment by CamiMargni** [comment]
@gastonprieto  El joinTable va para ambas herencias? Entiendo que las 5 intermedias son:
PubxTrago
CerveceríaXcerveza
BarxCerveza
BarxTrago
BarxBebidaSinAlcohol.

Entonces las one table de bebidas y locales estarían incorrectas?

**Comment by CamiMargni** [comment]
Enlazo el der como creo que quedaría
![Uploading image.png…]()

**Comment by gastonprieto** [comment]
> Buenas @gastonprieto , yo lo pense como los chicos de arriba y la verdad que le encontraba todo el sentido al DER y al modelo de objetos, no comprendo qué es lo que te limita usar una sola tabla intermedia si a fin y al cabo eso se delega en el ORM y basta con tener un ID de ambos lados como para saber a que registro especifico de cada tabla refiere. Perdon por la molestia, pero ¿podrias especificar un toque mas a que te referis con que no podria saber a cual corresponde?

Porque cuando estas haciendo el join contra la tabla intermedia no tenes información sobre el tipo particular de esa asociación, no sabes si es `Cerveza`,  `Analcoholica` o `Trago`

Entonces Hibernate no va a saber distinguir que objeto corresponde contra cual relación, recién haciendo un join contra la single table tal vez podría darse cuenta, pero eso es recién después de que intento joinear y joineo un monton de info separada. Y justo acá eso se puede hacer porque son disjuntas las colecciones, pero si compartieran algún tipo es imposible.

Entonces tiene que manejarse con colecciones separadas que representen las relaciones.

La ManyToMany esa da a entender en el modelo de datos que el Local tiene *una* lista de *Tragos*, lo cual no es correcto, son 5 listas que se manejan por separado.

El DER esta ok

**Comment by FeliIsra** [comment]
Dejo un ejemplo de Mappeo, por si alguien quiere comparar.

[practicaDDS.txt](https://github.com/dds-jv/foro/files/5599527/practicaDDS.txt)

**Comment by tatonico** [comment]
@gastonprieto Claro, yo habia entendido que en dominio de objetos se tenia como atributo una unica lista a una clase generica entre las 3 que seria Bebida, siendo 3 listas distintas comprendo lo que comentas. Por otro lado, es necesario el JoinTable y aclarar el nombre de la tabla intermedia?

**Comment by gastonprieto** [comment]
>  yo habia entendido que en dominio de objetos se tenia como atributo una unica lista a una clase generica entre las 3 que seria Bebida, siendo 3 listas distintas comprendo lo que comentas

Por las dudas son 5 listas, 3 + 1 + 1, y no, fijate que estan tipadas a las subclases

Tecnicamente es necesario, si en el parcial se lo olvidaban no pasaba nada.
